lang/rust/avro/src/types.rs

```diff
@@ -918,10 +922,25 @@ impl Value {
             Value::Long(n) => Ok(Value::Double(n as f64)),
             Value::Float(x) => Ok(Value::Double(f64::from(x))),
             Value::Double(x) => Ok(Value::Double(x)),
+            Value::String(x) => match Self::parse_special_float(&x) {
+                Some(f) => Ok(Value::Double(f.into())),
+                None => Err(Error::GetDouble(ValueKind::String)),
+            },
             other => Err(Error::GetDouble(other.into())),
         }
     }
 
+    /// IEEE 754 NaN and infinities are not valid JSON numbers.
+    /// So they are represented in JSON as strings.
+    fn parse_special_float(s: &str) -> Option<f32> {
+        match s.trim().to_ascii_lowercase().as_str() {
+            "nan" | "+nan" | "-nan" => Some(f32::NAN),
```

Comment:
[Java](https://github.com/apache/avro/pull/3066) and [C++](https://github.com/apache/avro/blob/release-1.12.0/lang/c%2B%2B/impl/parsing/JsonCodec.cc#L536) use case-sensitive values.
I suggest to compare against `NaN`, `Inf`, `Infinity`, etc., i.e. let's remove `to_asciilowercase()`